In [ ]:
import os
import pandas as pd
from pathlib import Path
import json
import re
from stereomapper.domain.chemistry.analysis import StereoAnalyser
from stereomapper.domain.chemistry import ChemistryUtils
from stereomapper.domain.chemistry import OpenBabelOperations
import sqlite3
import subprocess

In [ ]:
# run stereomapper as a subprocess
enantiomers = Path.home('enantiomer_benchmark_data')
results = Path.home('enantiomer_benchmark_results.sqlite') 
cache_path = Path.home('enantiomer_benchmark_cache.sqlite')

cmd =[
    "stereomapper",
    "run",
    "-d", enantiomers.as_posix(),
    "-o", results.as_posix(),
    "-p", cache_path.as_posix(),
    "--fresh-cache"
]

subprocess.run(cmd) 

INFO    Logging initialised. File: /media/JACK/stereomapper_logs/stereomapper_20251104_101611.log
Pipeline: 100%|██████████████████████████████████████████████████████████| , Complete! [00:07<00:00]


✅ Pipeline completed in 7.5s
📦 Inputs attempted: 2,504 (skipped 0)
📊 Successes: 2,500 | Failures: 4
🔗 Inchikey groups — processed 931, skipped 0, failed 0
🧮 Relationship rows: 3,618
🧾 Unique inchikeys observed: 931
💾 Cache hit rate: 0.0%


Pipeline: 100%|██████████████████████████████████████████████████████████| , Complete! [00:08<00:00]


CompletedProcess(args=['stereomapper', 'run', '-d', '/home/jackmcgoldrick/drive/Recon4IMD/WP5_Reconstruction/T5.1_ReconXKG/data/stereomapper/control_sets/stereoisomer_control/enantiomer_benchmark_data', '-o', '/home/jackmcgoldrick/drive/Recon4IMD/WP5_Reconstruction/T5.1_ReconXKG/results/stereomapper/project_results/benchmarking/041125_enantiomer_benchmark.sqlite', '-p', '/home/jackmcgoldrick/drive/Recon4IMD/WP5_Reconstruction/T5.1_ReconXKG/results/stereomapper/project_results/benchmarking/cache/041125_enantiomer_benchmark.sqlite', '--fresh-cache'], returncode=0)

In [44]:
conn = sqlite3.connect(results)
merged_df_q = """ 
SELECT * from relationships;
"""
df_merged = pd.read_sql_query(merged_df_q, conn)
df_merged

cluster_a  cluster_b       cluster_a_members       cluster_b_members  \
0             1          2   ["chebi:CHEBI:17521"]   ["chebi:CHEBI:36124"]   
1             3          4  ["chebi:CHEBI:137507"]  ["chebi:CHEBI:137513"]   
2             5          6  ["chebi:CHEBI:134198"]  ["chebi:CHEBI:134199"]   
3             7          8   ["chebi:CHEBI:28651"]   ["chebi:CHEBI:27702"]   
4             9         10  ["chebi:CHEBI:145480"]  ["chebi:CHEBI:145483"]   
...         ...        ...                     ...                     ...   
3613       2456       2458  ["chebi:CHEBI:133313"]  ["chebi:CHEBI:133312"]   
3614       2457       2458  ["chebi:CHEBI:133311"]  ["chebi:CHEBI:133312"]   
3615       2459       2460  ["chebi:CHEBI:233960"]  ["chebi:CHEBI:233959"]   
3616       2461       2462  ["chebi:CHEBI:142550"]    ["chebi:CHEBI:9399"]   
3617       2463       2464  ["chebi:CHEBI:138363"]  ["chebi:CHEBI:138362"]   

      cluster_a_size  cluster_b_size classification  score  \
0                  1               1  Diastereomers  100.0   
1                  1               1    Enantiomers  100.0   
2                  1               1    Enantiomers  100.0   
3                  1               1    Enantiomers  100.0   
4                  1               1    Enantiomers  100.0   
...              ...             ...            ...    ...   
3613               1               1  Diastereomers  100.0   
3614               1               1  Diastereomers  100.0   
3615               1               1    Enantiomers  100.0   
3616               1               1    Enantiomers  100.0   
3617               1               1    Enantiomers  100.0   

                  score_details extra_info version_tag  
0     {"confidence_bin":"high"}       None        v1.0  
1     {"confidence_bin":"high"}       None        v1.0  
2     {"confidence_bin":"high"}       None        v1.0  
3     {"confidence_bin":"high"}       None        v1.0  
4     {"confidence_bin":"high"}       None        v1.0  
...                         ...        ...         ...  
3613  {"confidence_bin":"high"}       None        v1.0  
3614  {"confidence_bin":"high"}       None        v1.0  
3615  {"confidence_bin":"high"}       None        v1.0  
3616  {"confidence_bin":"high"}       None        v1.0  
3617  {"confidence_bin":"high"}       None        v1.0  

[3618 rows x 11 columns]

In [52]:
expanded_keys = set()
for _, row in df_merged.iterrows():
    if not row["cluster_a_members"] or not row["cluster_b_members"]:
        continue
    m1s = [norm_chebi(x) for x in json.loads(row["cluster_a_members"])]
    m2s = [norm_chebi(x) for x in json.loads(row["cluster_b_members"])]
    for a in m1s:
        for b in m2s:
            if a == b:
                continue
            expanded_keys.add(tuple(sorted((a, b))))

# --- Overlap ---
tp = control_keys & expanded_keys
fp = expanded_keys - control_keys
fn = control_keys - expanded_keys

print(f"True Positives: {len(tp)}")
print(f"False Positives: {len(fp)}")
print(f"False Negatives: {len(fn)}")


True Positives: 1243
False Positives: 2494
False Negatives: 119


Not a true representation of the results, contains 2384 false positives majority of which are off target relationships. Therefore, the dataset needs to be remapped back onto the original pairs stored in the control dataset.

In [53]:
valid_pairs = control_keys

df_pred_filtered = []
for _, row in df_merged.iterrows():
    members_1 = [norm_chebi(x) for x in json.loads(row["cluster_a_members"])]
    members_2 = [norm_chebi(x) for x in json.loads(row["cluster_b_members"])]
    for a in members_1:
        for b in members_2:
            pair = tuple(sorted((a, b)))
            if pair in valid_pairs:
                df_pred_filtered.append({"id1": a, "id2": b, **row.to_dict()})

df_pred_filtered = pd.DataFrame(df_pred_filtered)

In [54]:
df_pred_filtered = df_pred_filtered.copy()
df_pred_filtered["pair_key"] = df_pred_filtered.apply(
    lambda r: tuple(sorted([norm_chebi(r["id1"]), norm_chebi(r["id2"])])), axis=1
)

In [55]:
df_pred_filtered

id1           id2  cluster_a  cluster_b  \
0      CHEBI_17521   CHEBI_36124          1          2   
1     CHEBI_137507  CHEBI_137513          3          4   
2     CHEBI_134198  CHEBI_134199          5          6   
3      CHEBI_28651   CHEBI_27702          7          8   
4     CHEBI_145480  CHEBI_145483          9         10   
...            ...           ...        ...        ...   
1238  CHEBI_133315  CHEBI_133312       2455       2458   
1239  CHEBI_133313  CHEBI_133311       2456       2457   
1240  CHEBI_233960  CHEBI_233959       2459       2460   
1241  CHEBI_142550    CHEBI_9399       2461       2462   
1242  CHEBI_138363  CHEBI_138362       2463       2464   

           cluster_a_members       cluster_b_members  cluster_a_size  \
0      ["chebi:CHEBI:17521"]   ["chebi:CHEBI:36124"]               1   
1     ["chebi:CHEBI:137507"]  ["chebi:CHEBI:137513"]               1   
2     ["chebi:CHEBI:134198"]  ["chebi:CHEBI:134199"]               1   
3      ["chebi:CHEBI:28651"]   ["chebi:CHEBI:27702"]               1   
4     ["chebi:CHEBI:145480"]  ["chebi:CHEBI:145483"]               1   
...                      ...                     ...             ...   
1238  ["chebi:CHEBI:133315"]  ["chebi:CHEBI:133312"]               1   
1239  ["chebi:CHEBI:133313"]  ["chebi:CHEBI:133311"]               1   
1240  ["chebi:CHEBI:233960"]  ["chebi:CHEBI:233959"]               1   
1241  ["chebi:CHEBI:142550"]    ["chebi:CHEBI:9399"]               1   
1242  ["chebi:CHEBI:138363"]  ["chebi:CHEBI:138362"]               1   

      cluster_b_size classification  score              score_details  \
0                  1  Diastereomers  100.0  {"confidence_bin":"high"}   
1                  1    Enantiomers  100.0  {"confidence_bin":"high"}   
2                  1    Enantiomers  100.0  {"confidence_bin":"high"}   
3                  1    Enantiomers  100.0  {"confidence_bin":"high"}   
4                  1    Enantiomers  100.0  {"confidence_bin":"high"}   
...              ...            ...    ...                        ...   
1238               1    Enantiomers  100.0  {"confidence_bin":"high"}   
1239               1    Enantiomers  100.0  {"confidence_bin":"high"}   
1240               1    Enantiomers  100.0  {"confidence_bin":"high"}   
1241               1    Enantiomers  100.0  {"confidence_bin":"high"}   
1242               1    Enantiomers  100.0  {"confidence_bin":"high"}   

     extra_info version_tag                      pair_key  
0          None        v1.0    (CHEBI_17521, CHEBI_36124)  
1          None        v1.0  (CHEBI_137507, CHEBI_137513)  
2          None        v1.0  (CHEBI_134198, CHEBI_134199)  
3          None        v1.0    (CHEBI_27702, CHEBI_28651)  
4          None        v1.0  (CHEBI_145480, CHEBI_145483)  
...         ...         ...                           ...  
1238       None        v1.0  (CHEBI_133312, CHEBI_133315)  
1239       None        v1.0  (CHEBI_133311, CHEBI_133313)  
1240       None        v1.0  (CHEBI_233959, CHEBI_233960)  
1241       None        v1.0    (CHEBI_142550, CHEBI_9399)  
1242       None        v1.0  (CHEBI_138362, CHEBI_138363)  

[1243 rows x 14 columns]

In [56]:
df_eval = pd.merge(
    dfc_dedup_wcs_excluded, df_pred_filtered,
    on="pair_key",
    how="left",
    suffixes=("_true", "_pred")
)

In [57]:
df_eval

id1_true      id2_true        label  \
0      CHEBI_43796   CHEBI_30314  Enantiomers   
1      CHEBI_15570   CHEBI_16977  Enantiomers   
2      CHEBI_17561   CHEBI_16375  Enantiomers   
3      CHEBI_17053   CHEBI_17364  Enantiomers   
4      CHEBI_29990   CHEBI_29991  Enantiomers   
...            ...           ...          ...   
1247  CHEBI_233695  CHEBI_176808  Enantiomers   
1248  CHEBI_233134  CHEBI_233952  Enantiomers   
1249  CHEBI_233954  CHEBI_233133  Enantiomers   
1250  CHEBI_233959  CHEBI_233960  Enantiomers   
1251   CHEBI_76640  CHEBI_195630  Enantiomers   

                                               molfile1  \
0     /media/JACK/stereomapper_structures/chebi_struc...   
1     /media/JACK/stereomapper_structures/chebi_struc...   
2     /media/JACK/stereomapper_structures/chebi_struc...   
3     /media/JACK/stereomapper_structures/chebi_struc...   
4     /media/JACK/stereomapper_structures/chebi_struc...   
...                                                 ...   
1247  /media/JACK/stereomapper_structures/chebi_struc...   
1248  /media/JACK/stereomapper_structures/chebi_struc...   
1249  /media/JACK/stereomapper_structures/chebi_struc...   
1250  /media/JACK/stereomapper_structures/chebi_struc...   
1251  /media/JACK/stereomapper_structures/chebi_struc...   

                                               molfile2         id1_c  \
0     /media/JACK/stereomapper_structures/chebi_struc...   CHEBI_43796   
1     /media/JACK/stereomapper_structures/chebi_struc...   CHEBI_15570   
2     /media/JACK/stereomapper_structures/chebi_struc...   CHEBI_17561   
3     /media/JACK/stereomapper_structures/chebi_struc...   CHEBI_17053   
4     /media/JACK/stereomapper_structures/chebi_struc...   CHEBI_29990   
...                                                 ...           ...   
1247  /media/JACK/stereomapper_structures/chebi_struc...  CHEBI_233695   
1248  /media/JACK/stereomapper_structures/chebi_struc...  CHEBI_233134   
1249  /media/JACK/stereomapper_structures/chebi_struc...  CHEBI_233954   
1250  /media/JACK/stereomapper_structures/chebi_struc...  CHEBI_233959   
1251  /media/JACK/stereomapper_structures/chebi_struc...   CHEBI_76640   

             id2_c                      pair_key      id1_pred      id2_pred  \
0      CHEBI_30314    (CHEBI_30314, CHEBI_43796)   CHEBI_30314   CHEBI_43796   
1      CHEBI_16977    (CHEBI_15570, CHEBI_16977)   CHEBI_15570   CHEBI_16977   
2      CHEBI_16375    (CHEBI_16375, CHEBI_17561)   CHEBI_17561   CHEBI_16375   
3      CHEBI_17364    (CHEBI_17053, CHEBI_17364)   CHEBI_17053   CHEBI_17364   
4      CHEBI_29991    (CHEBI_29990, CHEBI_29991)   CHEBI_29991   CHEBI_29990   
...            ...                           ...           ...           ...   
1247  CHEBI_176808  (CHEBI_176808, CHEBI_233695)  CHEBI_233695  CHEBI_176808   
1248  CHEBI_233952  (CHEBI_233134, CHEBI_233952)  CHEBI_233952  CHEBI_233134   
1249  CHEBI_233133  (CHEBI_233133, CHEBI_233954)  CHEBI_233954  CHEBI_233133   
1250  CHEBI_233960  (CHEBI_233959, CHEBI_233960)  CHEBI_233960  CHEBI_233959   
1251  CHEBI_195630   (CHEBI_195630, CHEBI_76640)  CHEBI_195630   CHEBI_76640   

      ...  cluster_b                           cluster_a_members  \
0     ...       34.0                       ["chebi:CHEBI:30314"]   
1     ...     1555.0                       ["chebi:CHEBI:15570"]   
2     ...     2239.0  ["chebi:CHEBI:17561", "chebi:CHEBI:35235"]   
3     ...      255.0                       ["chebi:CHEBI:17053"]   
4     ...      259.0                       ["chebi:CHEBI:29991"]   
...   ...        ...                                         ...   
1247  ...     2181.0                      ["chebi:CHEBI:233695"]   
1248  ...     2225.0                      ["chebi:CHEBI:233952"]   
1249  ...      929.0                      ["chebi:CHEBI:233954"]   
1250  ...     2460.0                      ["chebi:CHEBI:233960"]   
1251  ...     1427.0                      ["chebi:CHEBI:195630"]   

                               cluste

Dataframe `df_eval` now contains the original pairs of identifiers with their predicitions from stereomapper. This dataframe will be used to accurately calculate the precision, recall and F1 score on the enantiomer control dataset.

In [58]:
# only rows where we have a prediction - removing FN, ensure to state in results section
df_with_pred = df_eval.dropna(subset=["classification"])

tp = ((df_with_pred["classification"] == df_with_pred["label"]).sum())
fp = ((df_with_pred["classification"] != df_with_pred["label"]).sum())
fn = df_eval["classification"].isna().sum()

print(f"TP: {tp}, FP: {fp}, FN: {fn}")

TP: 1180, FP: 63, FN: 9


119 false negatives indicates rows where structures either could not be processed (due to errors in their structure files or issues with the parser compatibility) or they were excluded due to presence of wildcard atoms. 

In [59]:
precision = tp / (tp + fp) if (tp + fp) > 0 else 0
recall = tp / (tp + fn) if (tp + fn) > 0 else 0
f1 = 2 * precision * recall / (precision + recall) if (precision + recall) > 0 else 0

print(f"Precision: {precision:.3f}")
print(f"Recall:    {recall:.3f}")
print(f"F1 score:  {f1:.3f}")


Precision: 0.949
Recall:    0.992
F1 score:  0.970


Great results, lets investigate the false positive results to see what went wrong and if the pipeline can be improved.

In [77]:
df_fp = df_with_pred[df_with_pred["classification"] != df_with_pred["label"]]
df_fp

id1_true      id2_true        label  \
31     CHEBI_27372   CHEBI_27374  Enantiomers   
46     CHEBI_21397   CHEBI_21100  Enantiomers   
47     CHEBI_21101   CHEBI_21398  Enantiomers   
49     CHEBI_47537   CHEBI_16002  Enantiomers   
65     CHEBI_32922   CHEBI_35440  Enantiomers   
...            ...           ...          ...   
1235  CHEBI_231950  CHEBI_231949  Enantiomers   
1238  CHEBI_232288   CHEBI_76864  Enantiomers   
1240  CHEBI_232459  CHEBI_232353  Enantiomers   
1241  CHEBI_232532   CHEBI_76557  Enantiomers   
1251   CHEBI_76640  CHEBI_195630  Enantiomers   

                                               molfile1  \
31    /media/JACK/stereomapper_structures/chebi_struc...   
46    /media/JACK/stereomapper_structures/chebi_struc...   
47    /media/JACK/stereomapper_structures/chebi_struc...   
49    /media/JACK/stereomapper_structures/chebi_struc...   
65    /media/JACK/stereomapper_structures/chebi_struc...   
...                                                 ...   
1235  /media/JACK/stereomapper_structures/chebi_struc...   
1238  /media/JACK/stereomapper_structures/chebi_struc...   
1240  /media/JACK/stereomapper_structures/chebi_struc...   
1241  /media/JACK/stereomapper_structures/chebi_struc...   
1251  /media/JACK/stereomapper_structures/chebi_struc...   

                                               molfile2         id1_c  \
31    /media/JACK/stereomapper_structures/chebi_struc...   CHEBI_27372   
46    /media/JACK/stereomapper_structures/chebi_struc...   CHEBI_21397   
47    /media/JACK/stereomapper_structures/chebi_struc...   CHEBI_21101   
49    /media/JACK/stereomapper_structures/chebi_struc...   CHEBI_47537   
65    /media/JACK/stereomapper_structures/chebi_struc...   CHEBI_32922   
...                                                 ...           ...   
1235  /media/JACK/stereomapper_structures/chebi_struc...  CHEBI_231950   
1238  /media/JACK/stereomapper_structures/chebi_struc...  CHEBI_232288   
1240  /media/JACK/stereomapper_structures/chebi_struc...  CHEBI_232459   
1241  /media/JACK/stereomapper_structures/chebi_struc...  CHEBI_232532   
1251  /media/JACK/stereomapper_structures/chebi_struc...   CHEBI_76640   

             id2_c                      pair_key      id1_pred      id2_pred  \
31     CHEBI_27374    (CHEBI_27372, CHEBI_27374)   CHEBI_27374   CHEBI_27372   
46     CHEBI_21100    (CHEBI_21100, CHEBI_21397)   CHEBI_21100   CHEBI_21397   
47     CHEBI_21398    (CHEBI_21101, CHEBI_21398)   CHEBI_21101   CHEBI_21398   
49     CHEBI_16002    (CHEBI_16002, CHEBI_47537)   CHEBI_16002   CHEBI_47537   
65     CHEBI_35440    (CHEBI_32922, CHEBI_35440)   CHEBI_32922   CHEBI_35440   
...            ...                           ...           ...           ...   
1235  CHEBI_231949  (CHEBI_231949, CHEBI_231950)  CHEBI_231949  CHEBI_231950   
1238   CHEBI_76864   (CHEBI_232288, CHEBI_76864)   CHEBI_76864  CHEBI_232288   
1240  CHEBI_232353  (CHEBI_232353, CHEBI_232459)  CHEBI_232353  CHEBI_232459   
1241   CHEBI_76557   (CHEBI_232532, CHEBI_76557)   CHEBI_76557  CHEBI_232532   
1251  CHEBI_195630   (CHEBI_195630, CHEBI_76640)  CHEBI_195630   CHEBI_76640   

      ...  cluster_b       cluster_a_members       cluster_b_members  \
31    ...      233.0   ["chebi:CHEBI:27374"]   ["chebi:CHEBI:27372"]   
46    ...      417.0   ["chebi:CHEBI:21100"]   ["chebi:CHEBI:21397"]   
47    ...      409.0   ["chebi:CHEBI:21101"]   ["chebi:CHEBI:21398"]   
49    ...      404.0   ["chebi:CHEBI:16002"]   ["chebi:CHEBI:47537"]   
65    ...     1870.0   ["chebi:CHEBI:32922"]   ["chebi:CHEBI:35440"]   
...   ...        ...                     ...                     ...   
1235  ...     2229.0  ["chebi:CHEBI:231949"]  ["chebi:CHEBI:231950"]   
1238  ...      782.0   ["chebi:CHEBI:76864"]  ["chebi:CHEBI:232288"]   
1240  ...     2299.0  ["chebi:CHEBI:232353"]  ["chebi:CHEBI:232459"]   
1241  ...     1710.0   ["chebi:CHEBI:76557"]  ["chebi:CHEBI:232532"]   
1251  ...     1427.0  ["chebi:CHEBI:195630"]   ["chebi:CHEBI:7

In [70]:
df_fp['classification'].value_counts()

classification
Diastereomers    37
Parent-child     20
Unresolved        5
Unclassified      1
Name: count, dtype: int64

In [73]:
df_fp_dia = df_fp[df_fp['classification'] == "Diastereomers"]
df_fp_dia

id1_true      id2_true        label  \
31     CHEBI_27372   CHEBI_27374  Enantiomers   
46     CHEBI_21397   CHEBI_21100  Enantiomers   
47     CHEBI_21101   CHEBI_21398  Enantiomers   
49     CHEBI_47537   CHEBI_16002  Enantiomers   
110    CHEBI_18481   CHEBI_15387  Enantiomers   
164    CHEBI_28548   CHEBI_37209  Enantiomers   
175    CHEBI_37547   CHEBI_37546  Enantiomers   
232    CHEBI_38200   CHEBI_27669  Enantiomers   
266    CHEBI_17521   CHEBI_36124  Enantiomers   
278    CHEBI_46995   CHEBI_46987  Enantiomers   
282    CHEBI_47012   CHEBI_27476  Enantiomers   
408    CHEBI_50147   CHEBI_50146  Enantiomers   
425    CHEBI_17924   CHEBI_28789  Enantiomers   
449    CHEBI_50932   CHEBI_50931  Enantiomers   
595    CHEBI_64021   CHEBI_64020  Enantiomers   
609    CHEBI_67234   CHEBI_67233  Enantiomers   
933   CHEBI_132914  CHEBI_132913  Enantiomers   
934   CHEBI_132916  CHEBI_132915  Enantiomers   
1038  CHEBI_140637  CHEBI_136698  Enantiomers   
1039  CHEBI_140638  CHEBI_140636  Enantiomers   
1090  CHEBI_144940  CHEBI_138619  Enantiomers   
1153  CHEBI_156255   CHEBI_15414  Enantiomers   
1158  CHEBI_157606   CHEBI_72758  Enantiomers   
1169  CHEBI_167676  CHEBI_167675  Enantiomers   
1170  CHEBI_167883  CHEBI_167882  Enantiomers   
1171  CHEBI_167910  CHEBI_167901  Enantiomers   
1177  CHEBI_138641  CHEBI_138601  Enantiomers   
1186  CHEBI_183806   CHEBI_81467  Enantiomers   
1200  CHEBI_192698  CHEBI_134311  Enantiomers   
1206  CHEBI_193571   CHEBI_57463  Enantiomers   
1209  CHEBI_194352  CHEBI_194351  Enantiomers   
1211  CHEBI_195223  CHEBI_195225  Enantiomers   
1228  CHEBI_229701   CHEBI_57906  Enantiomers   
1232  CHEBI_231528   CHEBI_70980  Enantiomers   
1235  CHEBI_231950  CHEBI_231949  Enantiomers   
1238  CHEBI_232288   CHEBI_76864  Enantiomers   
1241  CHEBI_232532   CHEBI_76557  Enantiomers   

                                               molfile1  \
31    /media/JACK/stereomapper_structures/chebi_struc...   
46    /media/JACK/stereomapper_structures/chebi_struc...   
47    /media/JACK/stereomapper_structures/chebi_struc...   
49    /media/JACK/stereomapper_structures/chebi_struc...   
110   /media/JACK/stereomapper_structures/chebi_struc...   
164   /media/JACK/stereomapper_structures/chebi_struc...   
175   /media/JACK/stereomapper_structures/chebi_struc...   
232   /media/JACK/stereomapper_structures/chebi_struc...   
266   /media/JACK/stereomapper_structures/chebi_struc...   
278   /media/JACK/stereomapper_structures/chebi_struc...   
282   /media/JACK/stereomapper_structures/chebi_struc...   
408   /media/JACK/stereomapper_structures/chebi_struc...   
425   /media/JACK/stereomapper_structures/chebi_struc...   
449   /media/JACK/stereomapper_structures/chebi_struc...   
595   /media/JACK/stereomapper_structures/chebi_struc...   
609   /media/JACK/stereomapper_structures/chebi_struc...   
933   /media/JACK/stereomapper_structures/chebi_struc...   
934   /media/JACK/stereomapper_structures/chebi_struc...   
1038  /media/JACK/stereomapper_structures/chebi_struc...   
1039  /media/JACK/stereomapper_structures/chebi_struc...   
1090  /media/JACK/stereomapper_structures/chebi_struc...   
1153  /media/JACK/stereomapper_structures/chebi_struc...   
1158  /media/JACK/stereomapper_structures/chebi_struc...   
1169  /media/JACK/stereomapper_structures/chebi_struc...   
1170  /media/JACK/stereomapper_structures/chebi_struc...   
1171  /media/JACK/stereomapper_structures/chebi_struc...   
1177  /media/JACK/stereomapper_structures/chebi_struc...   
1186  /media/JACK/stereomapper_structures/chebi_struc...   
1200  /media/JACK/stereomapper_structures/chebi_struc...   
1206  /media/JACK/stereomapper_structures/chebi_struc...   
1209  /media/JACK/stereomapper_structures/chebi_struc...   
1211  /media/JACK/stereomapper_structures/chebi_struc...   
1228  /media/JACK/stereomapper_structures/chebi_struc...   
1232  /media/JACK/stereomapper_structures/chebi_struc...   
1235  /media/JACK/stereomapper_structures/chebi_

Lets investigate these cases manually to determine what went wrong.

Upon manual review, it has been determined that 23 out of the 37 pairs are deemed to be actual correct predictions by the stereomapper pipeline. An the 23 cases, each has been identified to be a mistake in assignment by ChEBI. Most of these relationships are indeed diastereomers, whilst others contain missing stereochemistry, which is picked up by the stereomapper pipeline. In one case, two structures are indeed enantiomers, but the two structures actually share different formal charges meaning they are not deemed enantiomers by the pipeline.

In reality, when accounting for these disagreements, stereomapper has the actual precision, recall and F1 score accounted for below:

In [75]:
df_fp_parent_child = df_fp[df_fp['classification'] == "Parent-child"]
df_fp_parent_child

id1_true      id2_true        label  \
65     CHEBI_32922   CHEBI_35440  Enantiomers   
172    CHEBI_37477   CHEBI_37476  Enantiomers   
210    CHEBI_38153   CHEBI_38154  Enantiomers   
237     CHEBI_3332   CHEBI_38969  Enantiomers   
254    CHEBI_39313   CHEBI_39309  Enantiomers   
256    CHEBI_39335   CHEBI_39336  Enantiomers   
588    CHEBI_63698   CHEBI_63695  Enantiomers   
589    CHEBI_63705   CHEBI_63700  Enantiomers   
705    CHEBI_77604   CHEBI_77599  Enantiomers   
777    CHEBI_83301   CHEBI_83300  Enantiomers   
885    CHEBI_90387   CHEBI_90386  Enantiomers   
886    CHEBI_90389   CHEBI_90391  Enantiomers   
887    CHEBI_90395   CHEBI_90394  Enantiomers   
994   CHEBI_137932   CHEBI_17426  Enantiomers   
1082  CHEBI_144384  CHEBI_144379  Enantiomers   
1192  CHEBI_189871  CHEBI_189872  Enantiomers   
1193  CHEBI_190007  CHEBI_190006  Enantiomers   
1223  CHEBI_228262  CHEBI_209277  Enantiomers   
1240  CHEBI_232459  CHEBI_232353  Enantiomers   
1251   CHEBI_76640  CHEBI_195630  Enantiomers   

                                               molfile1  \
65    /media/JACK/stereomapper_structures/chebi_struc...   
172   /media/JACK/stereomapper_structures/chebi_struc...   
210   /media/JACK/stereomapper_structures/chebi_struc...   
237   /media/JACK/stereomapper_structures/chebi_struc...   
254   /media/JACK/stereomapper_structures/chebi_struc...   
256   /media/JACK/stereomapper_structures/chebi_struc...   
588   /media/JACK/stereomapper_structures/chebi_struc...   
589   /media/JACK/stereomapper_structures/chebi_struc...   
705   /media/JACK/stereomapper_structures/chebi_struc...   
777   /media/JACK/stereomapper_structures/chebi_struc...   
885   /media/JACK/stereomapper_structures/chebi_struc...   
886   /media/JACK/stereomapper_structures/chebi_struc...   
887   /media/JACK/stereomapper_structures/chebi_struc...   
994   /media/JACK/stereomapper_structures/chebi_struc...   
1082  /media/JACK/stereomapper_structures/chebi_struc...   
1192  /media/JACK/stereomapper_structures/chebi_struc...   
1193  /media/JACK/stereomapper_structures/chebi_struc...   
1223  /media/JACK/stereomapper_structures/chebi_struc...   
1240  /media/JACK/stereomapper_structures/chebi_struc...   
1251  /media/JACK/stereomapper_structures/chebi_struc...   

                                               molfile2         id1_c  \
65    /media/JACK/stereomapper_structures/chebi_struc...   CHEBI_32922   
172   /media/JACK/stereomapper_structures/chebi_struc...   CHEBI_37477   
210   /media/JACK/stereomapper_structures/chebi_struc...   CHEBI_38153   
237   /media/JACK/stereomapper_structures/chebi_struc...    CHEBI_3332   
254   /media/JACK/stereomapper_structures/chebi_struc...   CHEBI_39313   
256   /media/JACK/stereomapper_structures/chebi_struc...   CHEBI_39335   
588   /media/JACK/stereomapper_structures/chebi_struc...   CHEBI_63698   
589   /media/JACK/stereomapper_structures/chebi_struc...   CHEBI_63705   
705   /media/JACK/stereomapper_structures/chebi_struc...   CHEBI_77604   
777   /media/JACK/stereomapper_structures/chebi_struc...   CHEBI_83301   
885   /media/JACK/stereomapper_structures/chebi_struc...   CHEBI_90387   
886   /media/JACK/stereomapper_structures/chebi_struc...   CHEBI_90389   
887   /media/JACK/stereomapper_structures/chebi_struc...   CHEBI_90395   
994   /media/JACK/stereomapper_structures/chebi_struc...  CHEBI_137932   
1082  /media/JACK/stereomapper_structures/chebi_struc...  CHEBI_144384   
1192  /media/JACK/stereomapper_structures/chebi_struc...  CHEBI_189871   
1193  /media/JACK/stereomapper_structures/chebi_struc...  CHEBI_190007   
1223  /media/JACK/stereomapper_structures/chebi_struc...  CHEBI_228262   
1240  /media/JACK/stereomapper_structures/chebi_struc...  CHEBI_232459   
1251  /media/JACK/stereomapper_structures/chebi_struc...   CHEBI_76640   

             id2_c                      pair_key      id1_pred      id2_pred  \
65     CHEBI_35440    (CHEBI_32922, CHEBI_35440)   CHEBI_32922   CHEBI_35440   
172   

All are incorrect predictions by StereoMapper.

In [76]:
df_fp_unresolved = df_fp[df_fp['classification'] == "Unresolved"]
df_fp_unresolved

id1_true     id2_true        label  \
168  CHEBI_15386  CHEBI_37465  Enantiomers   
208  CHEBI_38139  CHEBI_44343  Enantiomers   
281  CHEBI_47011  CHEBI_47008  Enantiomers   
763  CHEBI_83129  CHEBI_83131  Enantiomers   
764  CHEBI_83132  CHEBI_83130  Enantiomers   

                                              molfile1  \
168  /media/JACK/stereomapper_structures/chebi_struc...   
208  /media/JACK/stereomapper_structures/chebi_struc...   
281  /media/JACK/stereomapper_structures/chebi_struc...   
763  /media/JACK/stereomapper_structures/chebi_struc...   
764  /media/JACK/stereomapper_structures/chebi_struc...   

                                              molfile2        id1_c  \
168  /media/JACK/stereomapper_structures/chebi_struc...  CHEBI_15386   
208  /media/JACK/stereomapper_structures/chebi_struc...  CHEBI_38139   
281  /media/JACK/stereomapper_structures/chebi_struc...  CHEBI_47011   
763  /media/JACK/stereomapper_structures/chebi_struc...  CHEBI_83129   
764  /media/JACK/stereomapper_structures/chebi_struc...  CHEBI_83132   

           id2_c                    pair_key     id1_pred     id2_pred  ...  \
168  CHEBI_37465  (CHEBI_15386, CHEBI_37465)  CHEBI_37465  CHEBI_15386  ...   
208  CHEBI_44343  (CHEBI_38139, CHEBI_44343)  CHEBI_38139  CHEBI_44343  ...   
281  CHEBI_47008  (CHEBI_47008, CHEBI_47011)  CHEBI_47008  CHEBI_47011  ...   
763  CHEBI_83131  (CHEBI_83129, CHEBI_83131)  CHEBI_83131  CHEBI_83129  ...   
764  CHEBI_83130  (CHEBI_83130, CHEBI_83132)  CHEBI_83132  CHEBI_83130  ...   

     cluster_b      cluster_a_members      cluster_b_members cluster_a_size  \
168      503.0  ["chebi:CHEBI:37465"]  ["chebi:CHEBI:15386"]            1.0   
208     1566.0  ["chebi:CHEBI:38139"]  ["chebi:CHEBI:44343"]            1.0   
281     1780.0  ["chebi:CHEBI:47008"]  ["chebi:CHEBI:47011"]            1.0   
763     2221.0  ["chebi:CHEBI:83131"]  ["chebi:CHEBI:83129"]            1.0   
764     2219.0  ["chebi:CHEBI:83132"]  ["chebi:CHEBI:83130"]            1.0   

     cluster_b_size  classification  score                score_details  \
168             1.0      Unresolved  100.0    {"confidence_bin":"high"}   
208             1.0      Unresolved  100.0    {"confidence_bin":"high"}   
281             1.0      Unresolved  100.0    {"confidence_bin":"high"}   
763             1.0      Unresolved   75.0  {"confidence_bin":"medium"}   
764             1.0      Unresolved   75.0  {"confidence_bin":"medium"}   

                                            extra_info version_tag  
168  Possible pipeline error - should be no identic...        v1.0  
208  Possible pipeline error - should be no identic...        v1.0  
281  Possible pipeline error - should be no identic...        v1.0  
763  Possible pipeline error - should be no identic...        v1.0  
764  Possible pipeline error - should be no identic...        v1.0  

[5 rows x 21 columns]

all incorrect by StereoMapper. the single unclassified case is a complex case where stereochemistry differs, but so does protonation states.

In [78]:
# 23 for corrected FP in diastereomers, 1 for corrected FP in unclassified
tp_act = tp + 24
fp_act = fp - 24

actual_precision = tp_act / (tp_act + fp_act) if (tp_act + fp_act) > 0 else 0
actual_recall = tp_act / (tp_act + fn) if (tp_act + fn) > 0 else 0
actual_f1 = 2 * actual_precision * actual_recall / (actual_precision + actual_recall) if (actual_precision + actual_recall) > 0 else 0

print(f"Adjusted Precision: {actual_precision:.3f}")
print(f"Adjusted Recall:    {actual_recall:.3f}")
print(f"Adjusted F1 score:  {actual_f1:.3f}")

Adjusted Precision: 0.969
Adjusted Recall:    0.993
Adjusted F1 score:  0.980
